build(test): use shared library to reduce build artifacts

MattCatz opened the pull request:
Fixes #8127

Change the tests to use a shared library version of lvgl. This reduces the build artifact size on my machine by about 95%. Also enable `-Og` to enable compiler warning/checks that depend on optimization. `maybe-uninitialized`, `return-local-addr`, `stringop-overflow`, etc.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

kisvegabor commented:
Sorry that this PR was unanswered. Do you see any disadvantages of using an `so` file?

cc @liamHowatt @AndreCostaaa

AndreCostaaa commented:
Could this PR's fix be related to the CI problem? https://github.com/lvgl/lvgl/pull/8901

MattCatz commented:
> Ping @MattCatz , could you check the CI complaining? Thank you!

Yeah it really never left the draft state because it was never 100%. Since lvgl isn't typically built as a shared library, there are a few bugs in the CMake build files that only affect the shared build. i.e. build flags that should propagate but don't.

 I'll take another look at it.

MattCatz committed:
Fix overlapping glob

Since `THORVG_SOURCES` is under the `src/` directory, it is also
picked up by `SOURCES`. This means that liblvgl.a actually contains
the entirety of liblvgl_thorvg.a.

MattCatz committed:
Compile tests with a little bit of optimizations

Enable `-Og` to enable compiler warning/checks that depend on
optimization. `maybe-uninitialized`, `return-local-addr`,
`stringop-overflow`, etc.

MattCatz committed:
Add missing ASM flag for 32-bit tests

MattCatz committed:
Hidden symbols should be compiled inside the library.

MattCatz committed:
Build shared library for testing

MattCatz committed:
Link in missing library

MattCatz committed:
Increase tolerance in 32-bit tests

MattCatz commented:
@uLipe I did end up getting things working well enough. The shared library version doesn't work very well on windows (because of the circular dependencies) so I just dropped it (savings are mostly in the `test` actions not the `build` actions anyway). I also fixed a minor build bug with the source globs. The thorvg library is technically being built twice because it is being picked up by two different GLOBs. (and also always being included in `liblvgl.a`).

 I'm stumped on the last error. I think it's related to enabling optimizations on the 32-bit build.

It's unrelated but it would be nice if the checks didn't build `pngquant` every run. There is something funky in their build process that makes it take 10 minutes to build whereas locally it's near instantaneous.

AndreCostaaa committed:
build(test): add missing libdrm library link

AndreCostaaa committed:
chore: update ref images

AndreCostaaa commented:
Hi @MattCatz 

I also plan on doing some rework on how the tests are built and will try to add shared library support by then, let's warm this one up soon

Thank you for the work until now
